pkgs/dart_model/lib/src/lazy_merged_map.dart

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) 2024, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:collection';
+
+import 'dart_model.dart';
+
+/// An implementation of a lazy merged json [Map] view over two [Map]s.
+///
+/// The intended use case is for merging JSON payloads together into a single
+/// payload, where their structure is the same.
+///
+/// If both maps have the same key present, the logic for the values of those
+/// shared keys goes as follows:
+///
+///   - If both values are `Map<String, Object?>`, a nested [LazyMergedMapView]
+///     is returned.
+///   - Else if they are equal values, the value from [left] is returned.
+///   - Else a [StateError] is thrown.
+///
+/// Nested [List]s are not specifically handled at this time and must be equal.
+///
+/// The [keys] getter will de-duplicate the keys.
+class LazyMergedMapView extends MapBase<String, Object?> {
+  final Map<String, Object?> left;
+  final Map<String, Object?> right;
+
+  LazyMergedMapView(this.left, this.right);
+
+  @override
+  Object? operator [](Object? key) {
+    var leftValue = left[key];
```

Comment:
This can be pretty slow if we hit a case where both maps are buffer-backed and have large matching maps, since lookup is linear.

That's fine for now, there are lots of possibilities to optimize, we should do it with realistic use cases. Maybe a TODO for that?

Reply:
SGTM, I am not sure what we could do here exactly though, I think we always have to look up in each map.

Reply:
I was thinking more of optimized `values` and `entries` methods than optimized lookup. So it's still fast to just walk the whole `Map`.

Making lookup fast might already be a problem before merging :)

Anyway, something for later. Thanks.